File: app/src/main/java/com/github/fo2rist/mclaren/ui/widgets/InformationLineView.java
package com.github.fo2rist.mclaren.ui.widgets;


import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.fo2rist.mclaren.R;
import com.github.fo2rist.mclaren.utils.IntentUtils;
import com.github.fo2rist.mclaren.utils.LinkUtils;
import com.luseen.autolinklibrary.AutoLinkMode;
import com.luseen.autolinklibrary.AutoLinkOnClickListener;
import com.luseen.autolinklibrary.AutoLinkTextView;

/**
 * Single line with title/data content.
 */
public class InformationLineView extends LinearLayout implements AutoLinkOnClickListener {

    private TextView titleView;
    private AutoLinkTextView valueView;

    public InformationLineView(Context context) {
        this(context, (AttributeSet) null);
        this.setLayoutParams(getDefaultLayoutParams());
    }

    public InformationLineView(Context context, LinearLayout.LayoutParams layoutParams) {
        this(context, (AttributeSet) null);
        this.setLayoutParams(layoutParams);
    }

    public InformationLineView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.view_details_line, this);
        titleView = findViewById(R.id.title);
        valueView = findViewById(R.id.value);

        valueView.addAutoLinkMode(AutoLinkMode.MODE_MENTION);
        valueView.setAutoLinkOnClickListener(this);
        valueView.setMentionModeColor(ContextCompat.getColor(getContext(), R.color.textSecondaryWhite));
    }

    public void setContent(CharSequence title, CharSequence value) {
        titleView.setText(title);
        valueView.setAutoLinkText(value.toString());
    }

    @NonNull
    private LayoutParams getDefaultLayoutParams() {
        LayoutParams layoutParams = new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
        layoutParams.setMargins(
                getContext().getResources().getDimensionPixelSize(R.dimen.margin_five), //left
                getContext().getResources().getDimensionPixelSize(R.dimen.margin_half), //top
                getContext().getResources().getDimensionPixelSize(R.dimen.margin_five), //right
                getContext().getResources().getDimensionPixelSize(R.dimen.margin_half)  //bottom
        );
        return layoutParams;
    }

    @Override
    public void onAutoLinkTextClick(AutoLinkMode autoLinkMode, String autoLink) {
        IntentUtils.openInBrowser(getContext(), LinkUtils.getTwitterPageLink(autoLink));
    }
}

File: app/src/main/java/com/github/fo2rist/mclaren/ui/widgets/InformationLineView.kt
package com.github.fo2rist.mclaren.ui.widgets

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import com.github.fo2rist.mclaren.R
import com.github.fo2rist.mclaren.utils.IntentUtils
import com.github.fo2rist.mclaren.utils.LinkUtils
import io.github.armcha.autolink.AutoLinkItem
import io.github.armcha.autolink.AutoLinkTextView
import io.github.armcha.autolink.MODE_MENTION

/**
 * Single line with title/data content.
 */
class InformationLineView(
    context: Context,
    attrs: AttributeSet?
) : LinearLayout(context, attrs) {
    private var titleView: TextView
    private var valueView: AutoLinkTextView

    constructor(context: Context) : this(context, null as AttributeSet?) {
        this.layoutParams = createDefaultLayoutParams()
    }

    @Suppress("Unused")
    constructor(context: Context, layoutParams: LayoutParams) : this(context, null) {
        this.layoutParams = layoutParams
    }

    init {
        View.inflate(context, R.layout.view_details_line, this)
        titleView = findViewById(R.id.title)
        valueView = findViewById(R.id.value)
        valueView.addAutoLinkMode(MODE_MENTION)
        valueView.onAutoLinkClick(this::onAutoLinkTextClick);
        valueView.mentionModeColor = ContextCompat.getColor(context, R.color.textSecondaryWhite)
    }

    fun setContent(@StringRes title: Int, value: CharSequence) {
        titleView.setText(title)
        valueView.setText(value)
    }

    private fun createDefaultLayoutParams(): LayoutParams {
        return LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT).apply {
            setMargins(
                    context.resources.getDimensionPixelSize(R.dimen.margin_five),  //left
                    context.resources.getDimensionPixelSize(R.dimen.margin_half),  //top
                    context.resources.getDimensionPixelSize(R.dimen.margin_five),  //right
                    context.resources.getDimensionPixelSize(R.dimen.margin_half)   //bottom
            )
        }
    }

    private fun onAutoLinkTextClick(autoLink: AutoLinkItem) {
        IntentUtils.openInBrowser(context, LinkUtils.getTwitterPageLink(autoLink.originalText))
    }
}
